---
title: "Waterlogging microbiome analysis"
author: "Hamish McLean"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    number_sections: TRUE
    theme: cerulean
    highlight: tango
params:
  alpha_cache: TRUE
  pipeline: "DADA2"
---

## Setup

```{r setup, include=FALSE}

if (!exists("params")) params <- list(pipeline = "DADA2")

knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, error = TRUE)

# Kable options
# options(
#   browser = "false",
#   kableExtra.html.bsTable = TRUE,
#   kableExtra.bootstrap_options = c("striped", "hover", "condensed"),
#   kableExtra_view_html = interactive(),
#   knitr.table.format = "html",
#   viewer = NULL # Disable viewer to avoid opening in browser
# )

# Sys.setenv("BROWSER" = "false")

```

```{r libraries}

library(buildmer)
library(car)
library(cowplot)
library(data.table)
library(DESeq2)
library(dplyr)
library(DT)
library(emmeans)
library(ggpubr)
library(gridExtra)
library(here)
library(kableExtra)
library(lme4)
library(lmPerm)
library(partR2)
library(patchwork)
library(performance)
library(permutes)
library(see)
library(vegan)
library(viridis)

```

```{r functions}

here <- here::here()

FUNC_DIR <- here("functions")

source(here(FUNC_DIR, "functions.R"))
source(here(FUNC_DIR, "loadme.R"))
source(here(FUNC_DIR, "metabarcoding.R"))
source(here(FUNC_DIR, "metafuncs.R"))
source(here(FUNC_DIR, "rarefaction.R"))

# knit_print.datatable <- function(x, ...) {
#   htmltools::browsable(x)
# }

```

```{r constants}

DATA_DIR    <- normalizePath(here("..", "Data"))
FIGURES_DIR <- here("figures")
OBJECT_DIR  <- here("objects")

# Bioinformatics pipeline (defaults to DADA2)
# This should be set in the YAML header or passed as a parameter
PIPELINE    <- params$pipeline # DADA2 or USEARCH
ASV_FILTER  <- 0.001 # ASV filter for removing low abundance ASVs
READ_FILTER <- 0.05  # Read count filter for rarefaction
TAX_CONF    <- 0.65  # Confidence threshold for taxonomic rank assignment
SEED        <- 25647 # Random seed

alpha_cache <- params$alpha_cache

# Colour blind palette
cbPalette   <- c(
  "#000000", "#E69F00", "#56B4E9", "#009E73", 
  "#F0E442", "#0072B2", "#D55E00", "#CC79A7"
)

# Model designs
# Timepoint model for E1
T_DESIGN   <- ~ timepoint + block + treatment + timepoint:treatment + Error(plot)
T_DESIGN_M <- ~ timepoint * treatment + (1 | plot)

# Experiment model without timepoint 2
E_DESIGN   <- ~ experiment/block + treatment
E_DESIGN_M <- ~ experiment * treatment + (1 | experiment:block)

```

## Load data

Load data from R objects.

```{r load data}

FUN  <- readRDS(here(OBJECT_DIR, "FUN.rds"))
BAC  <- readRDS(here(OBJECT_DIR, "BAC.rds"))

EXP  <- readRDS(here(OBJECT_DIR, "EXP.rds"))
TIME <- readRDS(here(OBJECT_DIR, "TIME.rds"))

```

## Community size

### Copy number plots

```{r copy number plots}

copy_number_combined <- rbind(
  FUN$colData %>% mutate(type = "Fungi"),
  BAC$colData %>% mutate(type = "Bacteria")
)

copy_number_timepoint <- ggboxplot(
  filter(FUN$colData, experiment == 1), 
  x = "timepoint", y = "log10_copy_number", 
  fill = "weeks",
  xlab = "Timepoint",
  ylab = "log10 Copy number",
  facet.by = c("source")
)

copy_number_experiment <- ggboxplot(
  filter(FUN$colData, timepoint == 1), 
  x = "experiment", y = "log10_copy_number", 
  fill = "weeks",
  xlab = "Experiment",
  ylab = "log10 Copy number",
  facet.by = c("source")
)

cn_exp_plots <- ggarrange(
  copy_number_timepoint,
  copy_number_experiment,
  ncol = 1, nrow = 2,
  common.legend = TRUE, legend = "bottom"
)

ggsave(filename = "FUN_copy_number_plots.png", path = FIGURES_DIR)

cn_exp_plots


copy_number_timepoint <- ggboxplot(
  filter(BAC$colData, experiment == 1), 
  x = "timepoint", y = "log10_copy_number", 
  fill = "weeks",
  xlab = "Timepoint",
  ylab = "log10 Copy number",
  facet.by = c("source")
)

copy_number_experiment <- ggboxplot(
  filter(BAC$colData, timepoint == 1), 
  x = "experiment", y = "log10_copy_number", 
  fill = "weeks",
  xlab = "Experiment",
  ylab = "log10 Copy number",
  facet.by = c("source")
)

cn_time_plots <- ggarrange(
  copy_number_timepoint,
  copy_number_experiment,
  ncol = 1, nrow = 2,
  common.legend = TRUE, legend = "bottom"
)

ggsave(filename = "BAC_copy_number_plots.png", path = FIGURES_DIR)

cn_time_plots

```

### Copy number models

```{r copy number models}

exp_cn <- apply_to_subsets(EXP, copyNumberModelLMM, E_DESIGN_M) |> combine_tables()

time_cn <- apply_to_subsets(TIME, copyNumberModelLMM, T_DESIGN_M) |> combine_tables()

exp_cn[P <= 0.05 & factor != "(Intercept)"] |> printTable("Experiment")

time_cn[P <= 0.05 & factor != "(Intercept)"] |> printTable("Timepoint")

```

## Alpha diversity

### Calculate alpha diversity

Add alpha diversity idices to data objects.

```{r calculate alpha diversity}

EXP <- apply_to_subsets(EXP, calculateAlpha)
TIME <- apply_to_subsets(TIME, calculateAlpha)

```

### Permutation anova of alpha diversity index ranks

`perm.lmer` runs a permutation test for mixed-effects models.
This model takes a long time to run, so results are cached.

```{r alpha diversity analysis}

# These models are very slow, only run to update the cached results
if (!alpha_cache) {
  # Define metrics
  metrics <- c("S.chao1", "shannon", "simpson")

  exp_alpha  <- apply_to_subsets(EXP, alphaModel, metrics, E_DESIGN) |> combine_tables()
  time_alpha <- apply_to_subsets(TIME, alphaModel, metrics, T_DESIGN) |> combine_tables()

  saveRDS(exp_alpha,  here(OBJECT_DIR, "exp_alpha.rds"))
  saveRDS(time_alpha, here(OBJECT_DIR, "time_alpha.rds"))
} else {
  exp_alpha  <- readRDS(here(OBJECT_DIR, "exp_alpha.rds"))
  time_alpha <- readRDS(here(OBJECT_DIR, "time_alpha.rds"))
}

exp_alpha [p <= 0.05 & factor != "(Intercept)"] |> printTable()
time_alpha[p <= 0.05 & factor != "(Intercept)"] |> printTable()

```

#### Alpha post-hoc

```{r alpha post-hoc}

# EXP FUN root Simpson: NS, mean(simpson) slightly higher in treated for both
alphaPostHoc(EXP$FUN$root$alphaData, "simpson", formula(~ treatment + (1 | block)), "experiment")

# EXP FUN soil Chao1: NS, mean(Chao1) slighly higher in E1 trt, lower in E2 trt 
alphaPostHoc(EXP$FUN$soil$alphaData, "S.chao1", formula(~ treatment + (1 | block)), "experiment")

# EXP BAC soil Simpson: NS, mean(Simpson) slightly lower in treated for both
alphaPostHoc(EXP$BAC$soil$alphaData, "simpson", formula(~ treatment + (1 | block)), "experiment")



# TIME FUN soil Simpson: T1 ns, T2 *, higher in treated
alphaPostHoc(TIME$FUN$soil$alphaData, "simpson", formula(~ treatment + (1 | block)), "timepoint")

# TIME FUN soil Shannon: both ns, both lower in treated
alphaPostHoc(TIME$FUN$soil$alphaData, "shannon", formula(~ treatment + (1 | block)), "timepoint")

# TIME BAC root Chao1: both ns, 
alphaPostHoc(TIME$BAC$root$alphaData, "S.chao1", formula(~ treatment + (1 | block)), "timepoint")

# TIME BAC soil Simpson: T1 ns, T2 *, lower in treated
alphaPostHoc(TIME$BAC$soil$alphaData, "simpson", formula(~ treatment + (1 | block)), "timepoint")

```

### Alpha diversity plots

```{r alpha diversity plots}

measures <- c("Chao1", "Shannon", "Simpson")

plotAlpha <- function(data, k, s, measures, design) {
  plot_alpha(
    data$countData, data$colData,
    design = design, colour = "weeks",
    measures = measures,
    type = "box"
  ) + scale_colour_manual(values = cbPalette) +
    # theme(axis.title.x = element_blank()) +
    ggtitle(paste(k, ":", s))
}

exp_alpha_plots <- apply_to_subsets(EXP, plotAlpha, measures, "experiment") |> unlist(FALSE)

# ggarrange(exp_alpha_plots, ncol = 2, common.legend = TRUE, legend = "bottom")

# plot_grid(
#   plotlist = exp_alpha_plots,
#   align = "h",
#   ncol = 2
# )

(exp_alpha_plots[[1]] + exp_alpha_plots[[2]]) / (exp_alpha_plots[[3]] + exp_alpha_plots[[4]]) +
  plot_layout(guides = "collect") & # Collect legends into a single common legend
  theme(legend.position = "bottom") # Position the common legend

ggsave("alpha_EXP.png", path = FIGURES_DIR)


time_alpha_plots <- apply_to_subsets(TIME, plotAlpha, measures, "timepoint") |> unlist(FALSE)

(time_alpha_plots[[1]] + time_alpha_plots[[2]]) / (time_alpha_plots[[3]] + time_alpha_plots[[4]]) +
  plot_layout(guides = "collect") & # Collect legends into a single common legend
  theme(legend.position = "bottom") # Position the common legend

ggsave("alpha_TIME.png", path = FIGURES_DIR)

```

## Beta diversity: Bray-Curtis

### PERMANOVA of Bray-Curtis dissimilarity

Block included as a main effect in the design formula.

```{r adonis}

# Designs for ADONIS
# e_design <- vg ~ experiment/block * treatment

t_design <- vg ~ timepoint * treatment


# Apply adonis analysis to each subset of data
exp_bc  <- apply_to_subsets(EXP, adonisModel, update(E_DESIGN, vg ~ .)) |> combine_tables()

time_bc <- apply_to_subsets(TIME, adonisModel, t_design, "plot") |> combine_tables()

exp_bc[P < 0.05]  |> printTable()

time_bc[P < 0.05] |> printTable()

```

#### Post-hoc analysis

```{r adonis post-hoc}

#' @title Post-hoc analysis for adonis2 interaction terms
#' @description Splits data by a grouping variable and runs a simplified
#'   adonis2 model on each subset to investigate a significant interaction.
#' @param data A data.frame or data.table containing the experiment metadata.
#' @param counts A matrix or data.frame of raw or normalized counts (e.g., from DESeq2).
#' @param simplified_formula A formula object for the fixed effects of the simplified model (e.g., `~ treatment`).
#' @param group_by_var A string representing the column name to split the data by (e.g., "experiment" or "timepoint").
#' @param nperm The number of permutations to use for adonis2.
#' @return A list of the adonis2 results (data.tables) for each subset.
#' 
adonisPosthoc <- function(data, counts, simplified_formula, group_by_var, nperm = 1000) {

  # Split the data into a list of data.tables based on the grouping variable
  split_data <- split(data, data[[group_by_var]])

  # Initialize a list to store the results
  adonis_results <- list()

  # Iterate through each subset of the data
  for (group_name in names(split_data)) {
    subset_data <- split_data[[group_name]]

    # Subset the counts matrix to match the current metadata subset
    subset_counts <- counts[, rownames(subset_data)]

    # Calculate Bray-Curtis distances for the subset
    bray_curtis_dist <- vegan::vegdist(t(subset_counts), method = "bray")

    message(paste("Running adonis2 for", group_by_var, "=", group_name))

    # Define the full formula for the subset model
    # 'paste()' constructs new formula string like "bray_curtis_dist ~ treatment"
    full_formula_str <- paste("bray_curtis_dist", as.character(simplified_formula)[2], sep = " ")
    full_formula <- as.formula(full_formula_str)

    # Run adonis2 on the subset
    model_result <- tryCatch({
      vegan::adonis2(
        full_formula,
        data = subset_data,
        permutations = nperm,
        by = "terms"
      )
    }, error = function(e) {
      warning(paste("Error running adonis2 for", group_by_var, "=", group_name, ":", e$message))
      NULL
    })

    # Convert the adonis2 output to a data.frame for easier handling and store it
    if (!is.null(model_result)) {
      adonis_results[[group_name]] <- as.data.frame(model_result)
    }
  }

  # Return the list of adonis2 results
  return(adonis_results)
}

```

### Bray-Curtis NMDS plots

```{r nmds}

set.seed(SEED)

# Experiment

nmds_exp_fun_root <- nmdsPlot(EXP$FUN$root$dds, shape = "experiment", filename = "NMDS_FUN_root_exp.png")
nmds_exp_fun_soil <- nmdsPlot(EXP$FUN$soil$dds, shape = "experiment", filename = "NMDS_FUN_soil_exp.png")
nmds_exp_bac_root <- nmdsPlot(EXP$BAC$root$dds, shape = "experiment", filename = "NMDS_BAC_root_exp.png")
nmds_exp_bac_soil <- nmdsPlot(EXP$BAC$soil$dds, shape = "experiment", filename = "NMDS_BAC_soil_exp.png")

nmds_exp_combined <- ggarrange(
  plotlist = list(nmds_exp_fun_root, nmds_exp_fun_soil, nmds_exp_bac_root, nmds_exp_bac_soil),
  common.legend = TRUE,
  legend = "bottom",
  labels = c("Fungi: roots", "Fungi: soil", "Bacteria: roots", "Bacteria: soil")
)
nmds_exp_combined
ggsave("NMDS_exp.png", plot = nmds_exp_combined, path = FIGURES_DIR)

# Timepoint

nmds_time_fun_root <- nmdsPlot(TIME$FUN$root$dds, shape = "timepoint", filename = "NMDS_FUN_root_time.png")
nmds_time_fun_soil <- nmdsPlot(TIME$FUN$soil$dds, shape = "timepoint", filename = "NMDS_FUN_soil_time.png")
nmds_time_bac_root <- nmdsPlot(TIME$BAC$root$dds, shape = "timepoint", filename = "NMDS_BAC_root_time.png")
nmds_time_bac_soil <- nmdsPlot(TIME$BAC$soil$dds, shape = "timepoint", filename = "NMDS_BAC_soil_time.png")

nmds_time_combined <- ggarrange(
  plotlist = list(nmds_time_fun_root, nmds_time_fun_soil, nmds_time_bac_root, nmds_time_bac_soil),
  common.legend = TRUE,
  legend = "bottom",
  labels = c("Fungi: roots", "Fungi: soil", "Bacteria: roots", "Bacteria: soil")
)
nmds_time_combined
ggsave("NMDS_time.png", plot = nmds_time_combined, path = FIGURES_DIR)

```

## Beta diversity: PCA

### PCA models

```{r pca}

# Calculate PCA
EXP  <- apply_to_subsets(EXP, calculatePCA)
TIME <- apply_to_subsets(TIME, calculatePCA)


# EXP PC percent variance 
lapply(unlist(EXP, FALSE), function(x) {
  unlist(round(x$pca$percentVar * 100, 2))
}) |> as.data.table() |> mutate("PC" = 1:20) |> printTable()

# TIME PC percent variance
lapply(unlist(TIME, FALSE), function(x) {
  unlist(round(x$pca$percentVar * 100, 2))
}) |> as.data.table() |> mutate("PC" = 1:20) |> printTable()


# e_design <- ~ treatment * experiment + experiment:block
# t_design <- ~ treatment * timepoint

# EXP PCA linear model
apply_to_subsets(EXP, pcaLM, E_DESIGN) |> combine_tables() |> 
  filter(P <= 0.05 & factor != "(Intercept)") |> printTable()

# TIME PCA linear model
apply_to_subsets(TIME, pcaLM, T_DESIGN) |> combine_tables() |> 
  filter(P <= 0.05 & factor != "(Intercept)") |> printTable()


# EXP PCA linear mixed model
apply_to_subsets(EXP, pcaLMM, E_DESIGN_M) |> combine_tables() |> 
  filter(P <= 0.05 & factor != "(Intercept)") |> printTable()

# EXP PCA linear mixed model
apply_to_subsets(TIME, pcaLMM, T_DESIGN_M) |> combine_tables() |> 
  filter(P <= 0.05 & factor != "(Intercept)") |> printTable()

```

PCs of interest:

Experiment

- Fungi root: PC8 (int), PC9 (trt)

- Fungi soil: PC6 (trt, exp, & int)

- Bacteria root: PC7 (trt & int), PC8 (trt)

- Bacteria soil: PC3 (trt), PC4 (trt), PC8 (int), PC11 (trt), PC18 (int)

Timepoint

- Fungi root: PC4 (trt), PC5 (trt), PC10 (trt)

- Fungi soil: PC6 (int), PC7 (trt)

- Bacteria root: PC5 (trt, int), PC7 (trt, int), PC11 (int), PC12 (int)

- Bacteria soil: PC5 (trt, time, int), PC15(time, int)


### PC plots

```{r pc plots}

# EXP FUN root PCs 8 & 9
pcaPlot(EXP$FUN$root, c(8, 9), "experiment", "PCA_EXP_FUN_root_PC8-9.png")

# pcaPlot(EXP$FUN$soil, c(6), "experiment", "PCA_EXP_FUN_soil_PC6.png")

# EXP BAC root PCs 7 & 8
pcaPlot(EXP$BAC$root, c(7, 8), "experiment", "PCA_EXP_BAC_root_PC7-8.png")

# EXP BAC soil PCs 3 & 4
pcaPlot(EXP$BAC$soil, c(3, 4), "experiment", "PCA_EXP_BAC_soil_PC3-4.png")

# EXP BAC soil PCs 8 & 11
pcaPlot(EXP$BAC$soil, c(8, 11), "experiment", "PCA_EXP_BAC_soil_PC8-11.png")


# TIME FUN 

```

Insights

EXP FUN root PCs 8 & 9

EXP BAC root PCs 7 & 8
  PC7 exp 1 lest

### PC loadings

```{r pc loadings}

# PC loadings
pcLoadings <- function(data, pc, top_n = 10) {
  df <- cbind(PC = data$pca$x[, pc], data$colData)
  aggregate(PC ~ treatment, df, mean) |> print()

  loadings <- data$pca$rotation[, pc]
  top_asvs <- names(sort(abs(loadings), decreasing = TRUE))[1:top_n]
  data.table(
    PC = pc,
    ASVs = top_asvs,
    loading = loadings[top_asvs],
    taxa = data$taxData[top_asvs, ]$rank
  )
}

# Fungi root PC8
pcLoadings(EXP$FUN$root, 8, 10) |> printTable()

# Fungi root PC8
pcLoadings(EXP$FUN$root, 9, 10) |> printTable()

# Fungi soil PC6
pcLoadings(EXP$FUN$soil, 6, 10) |> printTable()


# Bacteria root PC7
pcLoadings(EXP$BAC$root, 7, 10) |> printTable()

# Bacteria soil PC4
pcLoadings(EXP$BAC$soil, 4, 10) |> printTable()

# Bacteria soil PC8
pcLoadings(EXP$BAC$soil, 8, 10) |> printTable()

```

Note: Multiply absolute loading by proportion of variance in PC explained by treatment
can then multiply this by proportion of variance of PC to the total community


## Differential abundance

```{r differential abundance}

# Run DESeq model
deseqModel <- function(data, k, s, design) {
  design(data$dds) <- design
  data$dds <- DESeq(data$dds)
  data
}

# e_design <- ~ treatment * experiment + experiment:block
# t_design <- ~ treatment * timepoint

EXP  <- apply_to_subsets(EXP, deseqModel, E_DESIGN)
TIME <- apply_to_subsets(TIME, deseqModel, T_DESIGN)

# Differential abundance

exp_da <- split(
  apply_to_subsets(EXP, diffAbundance, "experiment") |> combine_tables(),
  f = ~ kingdom
)

time_da <- split(
  apply_to_subsets(TIME, diffAbundance, "timepoint") |> combine_tables(),
  f = ~ kingdom
)

```

### Differential abundance tables {.tabset}

#### EXP: FUN

```{r}
exp_da$FUN |> printTable()
```

#### EXP: BAC

```{r}
exp_da$BAC |> printTable()
```

#### TIME: FUN

```{r}
time_da$FUN |> printTable()
```

#### TIME: BAC

```{r}
time_da$BAC |> printTable()
```

### Differential abundance subsets

```{r}

exp_da$FUN  |> select("taxa") |> table() |> as.data.table() |> printTable()
time_da$FUN |> select("taxa") |> table() |> as.data.table() |> printTable()

exp_da$FUN[grepl("Dactylonectria", taxa)] |> printTable()
time_da$FUN[grepl("Pleosporales", taxa)]  |> printTable()


exp_da$BAC  |> select("taxa") |> table() |> as.data.table() |> printTable()
time_da$BAC |> select("taxa") |> table() |> as.data.table() |> printTable()

time_da$BAC[grepl("Oxalobacteraceae", taxa)] |> printTable()

```

The fungal taxa with multiple occurences (but different ASVs) had wildly different
results.
There were fewer bacteria with multiple occurences, but the results were more consistant.
