---
title: "Waterlogging microbiome analysis"
author: "Hamish McLean"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    number_sections: TRUE
    theme: cerulean
    highlight: tango
params:
  alpha_cache: TRUE
  pipeline: "DADA2"
---

## Setup

```{r setup, include=FALSE}

if (!exists("params")) params <- list(pipeline = "DADA2")

knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, error = TRUE)

# Kable options
options(
  browser = "false",
  kableExtra.html.bsTable = TRUE,
  kableExtra.bootstrap_options = c("striped", "hover", "condensed"),
  kableExtra_view_html = interactive(),
  knitr.table.format = "html",
  viewer = NULL # Disable viewer to avoid opening in browser
)

Sys.setenv("BROWSER" = "false")

```

```{r libraries}

library(buildmer)
library(car)
library(data.table)
library(DESeq2)
library(dplyr)
library(emmeans)
library(ggpubr)
library(gridExtra)
library(here)
library(kableExtra)
library(lme4)
library(lmPerm)
library(partR2)
library(patchwork)
library(performance)
library(permutes)
library(see)
library(vegan)
library(viridis)

```

```{r functions}

here <- here::here()

FUNC_DIR <- here("functions")

source(here(FUNC_DIR, "functions.R"))
source(here(FUNC_DIR, "loadme.R"))
source(here(FUNC_DIR, "metabarcoding.R"))
source(here(FUNC_DIR, "metafuncs.R"))
source(here(FUNC_DIR, "rarefaction.R"))

```

```{r constants}

DATA_DIR    <- normalizePath(here("..", "Data"))
FIGURES_DIR <- here("figures")
OBJECT_DIR  <- here("objects")

# Bioinformatics pipeline (defaults to DADA2)
# This should be set in the YAML header or passed as a parameter
PIPELINE    <- params$pipeline # DADA2 or USEARCH
ASV_FILTER  <- 0.001 # ASV filter for removing low abundance ASVs
READ_FILTER <- 0.05  # Read count filter for rarefaction
TAX_CONF    <- 0.65  # Confidence threshold for taxonomic rank assignment
SEED        <- 25647 # Random seed

alpha_cache <- params$alpha_cache

# Colour blind palette
cbPalette   <- c(
  "#000000", "#E69F00", "#56B4E9", "#009E73", 
  "#F0E442", "#0072B2", "#D55E00", "#CC79A7"
)

# Model designs
# Timepoint model for E1
T_DESIGN <- ~ treatment * timepoint + (1 | plot)

# Experiment model without timepoint 2
E_DESIGN <- ~ treatment * experiment + (1 | experiment:block)

```

## Load data

Load data from R objects.

```{r load data}

FUN  <- readRDS(here(OBJECT_DIR, "FUN.rds"))
BAC  <- readRDS(here(OBJECT_DIR, "BAC.rds"))

EXP  <- readRDS(here(OBJECT_DIR, "EXP.rds"))
TIME <- readRDS(here(OBJECT_DIR, "TIME.rds"))

```

## Community size

Copy number plots

```{r copy number plots}

copy_number_combined <- rbind(
  FUN$colData %>% mutate(type = "Fungi"),
  BAC$colData %>% mutate(type = "Bacteria")
)

copy_number_timepoint <- ggboxplot(
  filter(FUN$colData, experiment == 1), 
  x = "timepoint", y = "log10_copy_number", 
  fill = "weeks",
  xlab = "Timepoint",
  ylab = "log10 Copy number",
  facet.by = c("source")
)

copy_number_experiment <- ggboxplot(
  filter(FUN$colData, timepoint == 1), 
  x = "experiment", y = "log10_copy_number", 
  fill = "weeks",
  xlab = "Experiment",
  ylab = "log10 Copy number",
  facet.by = c("source")
)

copy_number_plots <- ggarrange(
  copy_number_timepoint,
  copy_number_experiment,
  ncol = 1, nrow = 2,
  common.legend = TRUE, legend = "bottom"
)

ggsave(filename = "FUN_copy_number_plots.png", path = FIGURES_DIR)

copy_number_plots

copy_number_timepoint <- ggboxplot(
  filter(BAC$colData, experiment == 1), 
  x = "timepoint", y = "log10_copy_number", 
  fill = "weeks",
  xlab = "Timepoint",
  ylab = "log10 Copy number",
  facet.by = c("source")
)

copy_number_experiment <- ggboxplot(
  filter(BAC$colData, timepoint == 1), 
  x = "experiment", y = "log10_copy_number", 
  fill = "weeks",
  xlab = "Experiment",
  ylab = "log10 Copy number",
  facet.by = c("source")
)

ggarrange(
  copy_number_timepoint,
  copy_number_experiment,
  ncol = 1, nrow = 2,
  common.legend = TRUE, legend = "bottom"
)

ggsave(filename = "BAC_copy_number_plots.png", path = FIGURES_DIR)

```

Copy number models

```{r copy number models}

copy_number_model <- function(data, kingdom, source, design) {
  message <- ""
  model <- withCallingHandlers(
    lmer(update(design, copy_number ~ .), data$colData),
    message = function(m) {
      message <<- sub(":.*", "", conditionMessage(m))
      invokeRestart("muffleMessage")
    }
  )
  results <- Anova(model, type = 3) |> data.frame()
  data.table(
    kingdom = kingdom,
    source  = source,
    factor  = rownames(results),
    df      = results$Df,
    Chisq   = results$Chisq,
    P       = results$Pr..Chisq.,
    stars   = p_stars(results$Pr..Chisq.),
    message = message
  )
}

exp_cn <- apply_to_subsets(EXP, copy_number_model, E_DESIGN) |> combine_tables()

time_cn <- apply_to_subsets(TIME, copy_number_model, T_DESIGN) |> combine_tables()

exp_cn[P <= 0.05 & factor != "(Intercept)"]

time_cn[P <= 0.05 & factor != "(Intercept)"]

```

## Alpha diversity

### Calculate alpha diversity

```{r calculate alpha diversity}

#' @title Calculate alpha diversity
#' @description Calculate alpha diversity metrics for each subset of data
#' @param data A list containing countData and colData
#' @return A list with the alpha diversity data added
#' 
calculate_alpha_diversity <- function(data, ...) {
  alpha_dt <- plot_alpha(data$countData, data$colData, returnData = TRUE)
  data$alphaData <- alpha_dt[as.data.table(data$colData, keep.rownames = "Samples"), on = "Samples"]
  return(data)
}

EXP <- apply_to_subsets(EXP, calculate_alpha_diversity)
TIME <- apply_to_subsets(TIME, calculate_alpha_diversity)

```

### Permutation anova of alpha diversity index ranks

`perm.lmer` runs a permutation test for mixed-effects models.
This model takes a long time to run, so results are cached.

```{r alpha diversity analysis}

#' @title Analyze alpha diversity
#' @description Perform alpha diversity analysis on each subset of data
#' @param data A list containing countData and colData
#' @param kingdom A string representing the kingdom of the data (e.g., "Fungi" or "Bacteria")
#' @param source A string representing the source of the data (e.g., "root" or "soil")
#' @param metrics A vector of alpha diversity metrics to analyze
#' @param formula A formula for the model to use in the analysis
#' @return A list with the alpha diversity results
#'
analyse_alpha_diversity <- function(data, kingdom, source, metrics, formula) {
  cat(paste0("\nAnalyzing alpha diversity for ", kingdom, ": ", source, "\n"))

  # Run alpha diversity analysis for each metric
  anova_results <- list()
  for (met in metrics) {

    # Ensure the metric column exists
    if (!met %in% names(data$alphaData)) {
      warning("Metric '", met, "' not found in alpha data for ", kingdom, ": ", source)
      next
    }
    
    temp_dt <- copy(data$alphaData)
    temp_dt[, model_metric := rank(get(met))] # Rank of metric
    # Run the permutation model
    res <- tryCatch({
      perm.lmer(update(formula, model_metric ~ .), temp_dt, type = "anova", nperm = 1000)
    }, error = function(e) {
      warning("Error running perm.lmer for ", kingdom, ":", source, ", metric ", met, ": ", e$message)
      NULL # Return NULL on error
    })

    if (!is.null(res)) {
      res$metric <- met
      anova_results[[met]] <- res
    } else {
      anova_results[[met]] <- "Error" # Indicate error
      warning("ANOVA for ", kingdom, ":", source, ", metric ", met, " failed.")
    }
  }
  results <- rbindlist(anova_results)
  results$kingdom <- kingdom
  results$source  <- source
  results$stars   <- p_stars(results$p)
  results
}

# These models are very slow, only run to update the cached results
if (!alpha_cache) {
  # Define metrics
  metrics <- c("S.chao1", "shannon", "simpson")

  exp_alpha  <- apply_to_subsets(EXP, analyse_alpha_diversity, metrics, E_DESIGN) |> combine_tables()
  time_alpha <- apply_to_subsets(TIME, analyse_alpha_diversity, metrics, T_DESIGN) |> combine_tables()

  saveRDS(exp_alpha,  here(OBJECT_DIR, "exp_alpha.rds"))
  saveRDS(time_alpha, here(OBJECT_DIR, "time_alpha.rds"))
} else {
  exp_alpha <- readRDS(here(OBJECT_DIR, "exp_alpha.rds"))
  exp_alpha <- readRDS(here(OBJECT_DIR, "exp_alpha.rds"))
}

exp_alpha[p <= 0.05 & Factor != "(Intercept)"]
time_alpha[p <= 0.05 & Factor != "(Intercept)"]

```

### Alpha diversity plots

```{r alpha diversity plots}

measures <- c("Shannon", "Simpson")

# Root fungi

alpha_FUN_root_exp <- plot_alpha(
  EXP$FUN$root$countData, EXP$FUN$root$colData,
  design = "experiment", colour = "weeks",
  measures = measures,
  type = "box"
) + scale_colour_manual(values = cbPalette) +
  # theme(axis.title.x = element_blank()) +
  ggtitle("Fungi: root")
alpha_FUN_root_exp
ggsave(filename = "alpha_FUN_root_exp.png", path = FIGURES_DIR)

# Soil fungi

alpha_FUN_soil_exp <- plot_alpha(
  EXP$FUN$soil$countData, EXP$FUN$soil$colData,
  design = "experiment", colour = "weeks",
  measures = measures,
  type = "box"
) + scale_colour_manual(values = cbPalette) +
  # theme(axis.title.x = element_blank()) +
  ggtitle("Fungi: soil")
alpha_FUN_soil_exp
ggsave(filename = "alpha_FUN_soil_exp.png", path = FIGURES_DIR)

# Root bacteria

alpha_BAC_root_exp <- plot_alpha(
  EXP$BAC$root$countData, EXP$BAC$root$colData,
  design = "experiment", colour = "weeks",
  measures = measures,
  type = "box"
) + scale_colour_manual(values = cbPalette) +
  # theme(axis.title.x = element_blank()) +
  ggtitle("Bacteria: root")
alpha_BAC_root_exp
ggsave(filename = "alpha_BAC_root_exp.png", path = FIGURES_DIR)

# Soil bacteria

alpha_BAC_soil_exp <- plot_alpha(
  EXP$BAC$soil$countData, EXP$BAC$soil$colData,
  design = "experiment", colour = "weeks",
  measures = measures,
  type = "box"
) + scale_colour_manual(values = cbPalette) +
  # theme(axis.title.x = element_blank()) +
  ggtitle("Bacteria: soil")
alpha_BAC_soil_exp
ggsave(filename = "alpha_BAC_soil_exp.png", path = FIGURES_DIR)

# alpha_exp_combined <- ggarrange(
#   list(
#     alpha_FUN_root_exp, alpha_FUN_soil_exp,
#     alpha_BAC_root_exp, alpha_BAC_soil_exp
#   ),
#   ncol = 2, nrow = 2,
#   labels = c("Fungi Root", "Fungi Soil", "Bacteria Root", "Bacteria Soil"),
#   common.legend = TRUE, legend = "bottom"
# )

alpha_exp_combined <- (
  alpha_FUN_root_exp + alpha_FUN_soil_exp
) / (
  alpha_BAC_root_exp + alpha_BAC_soil_exp
) +
  plot_layout(guides = "collect") & # Collect legends into a single common legend
  theme(legend.position = "bottom") # Position the common legend


alpha_exp_combined
ggsave(
  filename = "alpha_exp_combined.png", plot = alpha_exp_combined, path = FIGURES_DIR
  # width = 10, height = 10, units = "cm"
)

```

## Beta diversity: Bray-Curtis

### PERMANOVA of Bray-Curtis dissimilarity

Block included as a main effect in the design formula.

```{r adonis}

e_design <- vg ~ treatment * experiment + experiment:block

t_design <- vg ~ treatment * timepoint

adonis_analysis <- function(data, kingdom, source, design, strata = NULL, nperm = 1000) {
  set.seed(SEED)
  vg <- vegdist(t(counts(data$dds, normalize = T)), method = "bray")
  model <- adonis2(
    update(design, vg ~ .),
    data$colData,
    strata = if (!is.null(strata)) data$colData[[strata]] else NULL,
    # strata = colData(dds)$experiment:colData(dds)$block, # For repeat measures
    permutations = nperm,
    by = "terms"
  )
  results <- data.frame(model)
  total_ss <- results["Total", "SumOfSqs"]
  data.table(
    kingdom = kingdom,
    source  = source,
    factor  = rownames(results),
    df      = results$Df,
    SS      = results$SumOfSqs,
    R2      = results$R2,
    F       = results$F,
    P       = results$Pr..F.,
    stars   = p_stars(results$Pr..F.),
    var     = results$SumOfSqs / total_ss * 100,
    n_perm  = attr(model, "control")$nperm
  )
}

adonis_analysis(EXP$FUN$soil, "FUN", "soil", e_design)

adonis_analysis(TIME$FUN$soil, "FUN", "soil", t_design, "plot")


# Apply adonis analysis to each subset of data
exp_bc  <- apply_to_subsets(EXP, adonis_analysis, e_design) |> combine_tables()

time_bc <- apply_to_subsets(TIME, adonis_analysis, t_design, "plot") |> combine_tables()

exp_bc[P < 0.05]

time_bc[P < 0.05]

```

### Bray-Curtis NMDS plots

```{r nmds}

set.seed(SEED)

#' @title NMDS plot
#' @description Create a NMDS plot for a given DESeq2 object
#' @param dds DESeq2 object
#' @param shape A string representing factor to use for the shape aesthetic
#' @param filename A string representing the filename to save the plot
#' @return A ggplot object of the NMDS plot
#'
nmdsPlot <- function(dds, shape, filename) {
  vg <- vegdist(t(counts(dds, normalize = T)), method = "bray")
  ord <- metaMDS(vg, trace=0)

  plot <- plotOrd(
    scores(ord), 
    colData(dds), 
    design = "treatment", 
    shape = shape, 
    alpha = 0.75, cbPalette = T
  )
  ggsave(filename, path = FIGURES_DIR)
  return(plot)
}

# Experiment

nmds_exp_fun_root <- nmdsPlot(EXP$FUN$root$dds, shape = "experiment", filename = "NMDS_FUN_root_exp.png")
nmds_exp_fun_soil <- nmdsPlot(EXP$FUN$soil$dds, shape = "experiment", filename = "NMDS_FUN_soil_exp.png")
nmds_exp_bac_root <- nmdsPlot(EXP$BAC$root$dds, shape = "experiment", filename = "NMDS_BAC_root_exp.png")
nmds_exp_bac_soil <- nmdsPlot(EXP$BAC$soil$dds, shape = "experiment", filename = "NMDS_BAC_soil_exp.png")

nmds_exp_combined <- ggarrange(
  plotlist = list(nmds_exp_fun_root, nmds_exp_fun_soil, nmds_exp_bac_root, nmds_exp_bac_soil),
  common.legend = TRUE,
  legend = "bottom",
  labels = c("Fungi: roots", "Fungi: soil", "Bacteria: roots", "Bacteria: soil")
)
nmds_exp_combined
ggsave("NMDS_exp.png", plot = nmds_exp_combined, path = FIGURES_DIR)

# Timepoint

nmds_time_fun_root <- nmdsPlot(TIME$FUN$root$dds, shape = "timepoint", filename = "NMDS_FUN_root_time.png")
nmds_time_fun_soil <- nmdsPlot(TIME$FUN$soil$dds, shape = "timepoint", filename = "NMDS_FUN_soil_time.png")
nmds_time_bac_root <- nmdsPlot(TIME$BAC$root$dds, shape = "timepoint", filename = "NMDS_BAC_root_time.png")
nmds_time_bac_soil <- nmdsPlot(TIME$BAC$soil$dds, shape = "timepoint", filename = "NMDS_BAC_soil_time.png")

nmds_time_combined <- ggarrange(
  plotlist = list(nmds_time_fun_root, nmds_time_fun_soil, nmds_time_bac_root, nmds_time_bac_soil),
  common.legend = TRUE,
  legend = "bottom",
  labels = c("Fungi: roots", "Fungi: soil", "Bacteria: roots", "Bacteria: soil")
)
nmds_time_combined
ggsave("NMDS_time.png", plot = nmds_time_combined, path = FIGURES_DIR)

```

## Beta diversity: PCA

### PCA models

```{r pca}


# fun_pca <- des_to_pca(FUN$dds)
# # fun_pca <- t(data.frame(t(pca$x) * pca$percentVar))

# bac_pca <- des_to_pca(BAC$dds)
# # bac_pca <- t(data.frame(t(pca$x) * pca$percentVar))

# Apply PCA to subsets
pca_subset <- function(data, name, source, n_pcs) {
  # Calculate PCA scores
  pca <- des_to_pca(data$dds)

  # Print PCA percent variance table
  pca_var <- data.frame(
    PC = paste0("PC", 1:n_pcs),
    perc_var = round(pca$percentVar[1:n_pcs] * 100, 1)
  )
  cat("\n", name, ":", source, "\n\n")
  print(pca_var)

  # Adjust PCA scores with total variance
  data$pca <- pca #t(data.frame(t(pca$x) * pca$percentVar))
  return(data)
}

n_pcs <- 10

EXP  <- apply_to_subsets(EXP, pca_subset, n_pcs)
TIME <- apply_to_subsets(TIME, pca_subset, n_pcs)


# Summarise PCA
pca_summary_subset <- function(data, kingdom, source, design, n_pcs = NULL) {
  pcs      <- 1:ifelse(!is.null(n_pcs), n_pcs, ncol(data$pca$x) - 1)
  pca_data <- as.data.frame(data$pca$x[, pcs])
  variance <- data$pca$percentVar[pcs]
  pc_names <- colnames(pca_data)

  results  <- lapply(seq_along(pcs), function(i) {
    pc_name  <- pc_names[i]
    df       <- cbind(PC = pca_data[[pc_name]], data$colData)
    model    <- lm(update(design, PC ~ .), df)
    anova    <- Anova(model, type = 3)
    anova_df <- anova |> data.frame()
    total_ss <- sum(anova_df$Sum.Sq)
    data.table(
      kingdom = kingdom,
      source  = source,
      pc      = pc_name,
      factor  = rownames(anova_df),
      SS      = anova_df$Sum.Sq,
      df      = anova_df$Df,
      F       = anova_df$F.value,
      P       = anova_df$Pr..F.,
      pc_var  = variance[i],
      var     = anova_df$Sum.Sq / total_ss * 100,
      var_adj = (anova_df$Sum.Sq / total_ss * 100) * variance[i]
    )
  })

  bind_rows(results)[P <= 0.05 & factor != "(Intercept)"]
}

# pca_summary_subset(EXP$FUN$soil, "FUN", "soil", e_design)

e_design <- ~ treatment * experiment + experiment:block
t_design <- ~ treatment * timepoint

apply_to_subsets(EXP, pca_summary_subset, e_design)  |> combine_tables()
apply_to_subsets(TIME, pca_summary_subset, t_design) |> combine_tables()


# Summarise PCA lmer
pca_summary_subset_lmer <- function(data, kingdom, source, design, n_pcs = NULL) {
  pcs      <- 1:ifelse(!is.null(n_pcs), n_pcs, ncol(data$pca$x) - 1)
  pca_data <- as.data.frame(data$pca$x[, pcs])
  pc_names <- colnames(pca_data)
  variance <- data$pca$percentVar[pcs]

  results <- lapply(seq_along(pcs), function(i) {
    pc_name  <- pc_names[i]
    df       <- cbind(PC = pca_data[[pc_name]], data$colData)
    message  <- NULL
    model    <- withCallingHandlers(
      lmer(update(design, PC ~ .), df),
      message = function(m) {
        msg <- conditionMessage(m)
        message <<- sub(":.*", "", msg)
        invokeRestart("muffleMessage")
      }
    )
    anova    <- Anova(model, type = 3) |> data.frame()
    # r2       <- r2_nakagawa(model)
    # part_r2 <- partR2(model, partvars = c("treatment", "timepoint", "treatment:timepoint"))
    # print(part_r2)
    data.table(
      kingdom = kingdom,
      source  = source,
      pc      = pc_name,
      factor  = rownames(anova),
      Chisq   = anova$Chisq,
      df      = anova$Df,
      P       = anova$Pr..Chisq.,
      # R2_marg = r2$marginal,
      # R2_cond = r2$conditional,
      # R2_part = part_r2$R2,
      pc_var  = variance[i],
      message = message
    )
  })

  bind_rows(results)[P <= 0.05 & factor != "(Intercept)"]
}

apply_to_subsets(EXP, pca_summary_subset_lmer, E_DESIGN)  |> combine_tables()

apply_to_subsets(TIME, pca_summary_subset_lmer, T_DESIGN) |> combine_tables()

```

PCs of interest:

Experiment
- Fungi root: PC8 (int), PC9 (trt)
- Fungi soil: PC6 (trt, exp, & int)
- Bacteria root: PC7 (trt & int), PC8 (trt)
- Bacteria soil: PC3 (trt), PC4 (trt), PC8 (int), PC11 (trt), PC18 (int)

Timepoint
- Fungi root: PC4 (trt), PC5 (trt), PC10 (trt)
- Fungi soil: PC6 (int), PC7 (trt)
- Bacteria root: PC5 (trt, int), PC7 (trt, int), PC11 (int), PC12 (int)
- Bacteria soil: PC5 (trt, time, int), PC15(time, int)

### PC plots

```{r pc plots}

#' @title PCA plot
#' @description Create a PCA plot for a given pair of PCs
#' @param data PCA data
#' @param shape A string representing factor to use for the shape aesthetic
#' @param filename A string representing the filename to save the plot
#' @return A ggplot object of the PCA plot
#'
pcaPlot <- function(data, pcs, shape, filename) {
  plot <- plotOrd(
    t(data.frame(t(data$pca$x) * data$pca$percentVar)),
    # data$pca, 
    data$colData, 
    design = "treatment", 
    shapes = shape, 
    axes = pcs,
    alpha = 0.75, 
    cbPalette = TRUE
  )
  ggsave(filename, path = FIGURES_DIR)
  return(plot)
}

# EXP FUN root PCs 8 & 9
pcaPlot(EXP$FUN$root, c(8, 9), "experiment", "PCA_EXP_FUN_root_PC8-9.png")

# pcaPlot(EXP$FUN$soil, c(6), "experiment", "PCA_EXP_FUN_soil_PC6.png")

# EXP BAC root PCs 7 & 8
pcaPlot(EXP$BAC$root, c(7, 8), "experiment", "PCA_EXP_BAC_root_PC7-8.png")

# EXP BAC soil PCs 3 & 4
pcaPlot(EXP$BAC$soil, c(3, 4), "experiment", "PCA_EXP_BAC_soil_PC3-4.png")

# EXP BAC soil PCs 8 & 11
pcaPlot(EXP$BAC$soil, c(8, 11), "experiment", "PCA_EXP_BAC_soil_PC8-11.png")


# TIME FUN 

```

Insights

EXP FUN root PCs 8 & 9

EXP BAC root PCs 7 & 8
  PC7 exp 1 lest

### PC loadings

```{r pc loadings}

# PC loadings
pc_loadings <- function(data, pc, top_n = 10) {
  df <- cbind(PC = data$pca$x[, pc], data$colData)
  aggregate(PC ~ treatment, df, mean) |> print()

  loadings <- data$pca$rotation[, pc]
  top_asvs <- names(sort(abs(loadings), decreasing = TRUE))[1:top_n]
  data.table(
    PC = pc,
    ASVs = top_asvs,
    loading = loadings[top_asvs],
    taxa = data$taxData[top_asvs, ]$rank
  )
}

# Fungi root PC8
pc_loadings(EXP$FUN$root, 8, 10)

# Fungi root PC8
pc_loadings(EXP$FUN$root, 9, 10)

# Fungi soil PC6
pc_loadings(EXP$FUN$soil, 6, 10)


# Bacteria root PC7
pc_loadings(EXP$BAC$root, 7, 10)

# Bacteria soil PC4
pc_loadings(EXP$BAC$soil, 4, 10)

# Bacteria soil PC8
pc_loadings(EXP$BAC$soil, 8, 10)

```

Note: Multiply absolute loading by proportion of variance in PC explained by treatment
can then multiply this by proportion of variance of PC to the total community


### Differential abundance

```{r differential abundance}

# Run DESeq model
deseq_model <- function(data, k, s, design) {
  design(data$dds) <- design
  data$dds <- DESeq(data$dds)
  data
}

e_design <- ~ treatment * experiment + experiment:block
t_design <- ~ treatment * timepoint

EXP  <- apply_to_subsets(EXP, deseq_model, e_design)
TIME <- apply_to_subsets(TIME, deseq_model, t_design)


# DESeq differential abundance analysis
diff_abundance <- function(data, kingdom, source, term) {
  int_term <- paste0("treatment1.", term, "2")
  # Extract results for E1 (reference), E2, and interaction
  res_x1  <- results(data$dds, name = "treatment_1_vs_0")
  res_x2  <- results(data$dds, contrast = list(c("treatment_1_vs_0", int_term)))
  res_int <- results(data$dds, name = int_term)

  # Combine into a data.table
  dt <- data.table(
    kingdom = kingdom,
    source = source,
    ASV = rownames(res_x1),
    log2FC_X1 = res_x1$log2FoldChange,
    padj_X1 = res_x1$padj,
    sig_X1 = p_stars(res_x1$padj),
    log2FC_X2 = res_x2$log2FoldChange,
    padj_X2 = res_x2$padj,
    sig_X2 = p_stars(res_x2$padj),
    log2FC_inter = res_int$log2FoldChange,
    padj_inter = res_int$padj,
    sig_inter = p_stars(res_int$padj),
    taxa = data$taxData[rownames(res_x1), ]$rank
  )
  
  # Filter for any significant result
  dt_filt <- dt[
    complete.cases(padj_X1, padj_X2, padj_inter) &
    pmin(padj_X1, padj_X2, padj_inter, na.rm = TRUE) < 0.05
  ]

  # Optionally, sort by log2FC
  dt_filt <- dt_filt[order(pmax(abs(log2FC_X1), abs(log2FC_X2), na.rm = TRUE), decreasing = TRUE)]
  
  return(dt_filt)
}

exp_da <- split(
  apply_to_subsets(EXP, diff_abundance, "experiment") |> combine_tables(),
  f = ~ kingdom
)

time_da <- split(
  apply_to_subsets(TIME, diff_abundance, "timepoint") |> combine_tables(),
  f = ~ kingdom
)

exp_da$FUN |> select("taxa") |> table() |> as.data.table()
time_da$FUN |> select("taxa") |> table() |> as.data.table()

exp_da$FUN[grepl("Dactylonectria", taxa)]
time_da$FUN[grepl("Pleosporales", taxa)]


exp_da$BAC |> select("taxa") |> table() |> as.data.table()
time_da$BAC |> select("taxa") |> table() |> as.data.table()

time_da$BAC[grepl("Oxalobacteraceae", taxa)]

```

The fungal taxa with multiple occurences (but different ASVs) had wildly different
results.
There were fewer bacteria with multiple occurences, but the results were more consistant.
